woocommerce/woocommerce
# Create and activate product attributes lookup table in data migration

### All Submissions:


### Changes proposed in this Pull Request:

Until now the product attributes lookup table had to be created and its usage had to be activated manually, via the tools page. This pull request does the following changes:

1. Remove the tools to create and delete the table (but keep the tool to regenerate the table data for one or for all the products).

2. Add a data migration that triggers the table creation and filling (unless the table already existed, in this case nothing happens).

3. After the migration finishes, activate the table usage for product filtering (site admin can still disable it via _Settings - Products - Advanced_); again, if the table already existed, nothing happens (the setting is unchanged). The setting will be left disabled if the regeneration is abored manually (see below).

Additionally, the following has been added in order to help in the migration process:

1. A new filter, `woocommerce_attribute_lookup_regeneration_step_size`, that allows to alter the batch size for the process (the number of products that will be processed each time that Action Scheduler runs one migration step). This filter is run on every step, thus it can be used to alter the batch size even if the regeneration process has been already started.

2. If the regeneration process is aborted manually (by deleting the `woocommerce_attribute_lookup_regeneration_in_progress` option), the _Products - Advanced_ settings section will still allow to enable the lookup table usage, but will display a warning discouraging to do so because the lookup table will be in an inconsistent state. The warning is shown when the `woocommerce_attribute_lookup_regeneration_aborted` option exists, and the option is deleted when the table regeneration is triggered again via the tools page.

![image](https://user-images.githubusercontent.com/937723/149765098-830d9792-e6b1-46e0-b9fa-6158a934c557.png)


Closes #30564.
Closes #30565.

### How to test the changes in this Pull Request:

1. If you had tested the product attributes lookup table mechanism already and the table still exists in your db, delete it from _Status - Tools - Delete the product attributes lookup table_ (you need to do this from `trunk`).

2. Switch to the branch of this pull request and go to _Status - Tools_. You shouldn't see any entry related to the product attributes lookup table.

3. Trigger the WooCommerce db update from the update available warning in the admin area.

3. After the update finishes you'll see that the old tools page entry "Regenerate the product attributes lookup table" is back. If you have a lot of products in your db the "Regenerate" button will be disabled and will show a counter of processed products, this means that the table regeneration is still in progress.

4. Once the table filling is complete the "Regenerate" product should appear as enabled. After that has happened go to _Settings - Products - Advanced_ and verify that "Enable table usage" is set.

5. Verify that the feature still works as expected (enabling and disabling the table usage via settings, regenerating the table from tools page for one product or for all products).

Now, to test the case of a db update when the table already exists:

6. Disable the table usage via settings.

7. Run the following in command line: `wp option set woocommerce_db_version 6.0.0`

8. Trigger the db update from the banner again, this time nothing related to the lookup table happens (the "Regenerate" tool has its button enabled right after the db update finishes).

9. Verify that the table usage is still disabled in settings.

To test the new filter:

10. Add the following somewhere (e.g. at the end of `woocommerce.php`):

```php
add_filter('woocommerce_attribute_lookup_regeneration_step_size', function() {
	return 5;
});
```

Verify that the regeneration happens now in increments of five products (hint: use `wp option get woocommerce_attribute_lookup_processed_count` to easily get the count of products already processed). If you change the value mid-regeneration you should see that the new value is used from the following regeneration step.

To test the new warning message:

11. Abort an already started regeneration by running `wp option delete woocommerce_attribute_lookup_regeneration_in_progress`.

12. Verify that the _Products - Advanced_ settings page now shows the warning (you may need to reload a couple of times to give Action Scheduler a chance to run one last time).

13. Start a new regeneration process, let it finish, and verify that the warning has disappeared.

### Other information:




### Changelog entry

> Add - Data migration to create and activate the product attributes lookup table.

### FOR PR REVIEWER ONLY:
